src/java.desktop/share/classes/javax/swing/JEditorPane.java

```diff
@@ -472,69 +472,82 @@ public void setPage(URL page) throws IOException {
         if (page == null) {
             throw new IOException("invalid url");
         }
-        URL loaded = getPage();
 
+        final URL loaded = getPage();
+        final Object postData = getPostData();
+        final String reference = page.getRef();
 
-        // reset scrollbar
-        if (!page.equals(loaded) && page.getRef() == null) {
-            scrollRectToVisible(new Rectangle(0,0,1,1));
-        }
-        boolean reloaded = false;
-        Object postData = getPostData();
-        if ((loaded == null) || !loaded.sameFile(page) || (postData != null)) {
-            // different url or POST method, load the new content
-
-            int p = getAsynchronousLoadPriority(getDocument());
-            if (p < 0) {
-                // open stream synchronously
-                InputStream in = getStream(page);
-                if (kit != null) {
-                    Document doc = initializeModel(kit, page);
-
-                    // At this point, one could either load up the model with no
-                    // view notifications slowing it down (i.e. best synchronous
-                    // behavior) or set the model and start to feed it on a separate
-                    // thread (best asynchronous behavior).
-                    p = getAsynchronousLoadPriority(doc);
-                    if (p >= 0) {
-                        // load asynchronously
-                        setDocument(doc);
-                        synchronized(this) {
-                            pageLoader = new PageLoader(doc, in, loaded, page);
-                            pageLoader.execute();
-                        }
-                        return;
-                    }
-                    read(in, doc);
-                    setDocument(doc);
-                    reloaded = true;
-                }
+        if ((postData == null) && page.sameFile(loaded)) {
+            // The same page with different reference
+            if (reference != null) {
+                scrollToReference(reference);
```

Comment:
Do we need to move ` scrollToReference(reference);` to swingUtil ? If not any reason for invoking it in swingUtil at line 547 and not here?

Reply:
Is it because of same page which is already loaded and no reloading is required?

Reply:
Yes, the page is already loaded, it's rendered, so the location to scroll to can be calculated right away.

Below, there's a comment:
https://github.com/openjdk/jdk/blob/be168e9d331ead58503ac3a01dae0bd98ae42d5e/src/java.desktop/share/classes/javax/swing/JEditorPane.java#L544-L549

This piece of code is unchanged. The new document is just loaded. To calculate the location to scroll to, the view hierarchy needs to be created and laid out; this is why `invokeLater` is used — to delay calculating the position.